test/sql/test_types.py

```diff
@@ -383,6 +384,28 @@ def load_dialect_impl(self, dialect):
         assert t1.dialect_impl(d)._type_affinity is postgresql.UUID
 
 
+class AsGenericTest(fixtures.TestBase):
+    @testing.combinations(
+        (String(), String()),
+        (VARCHAR(length=100), String(length=100)),
+        (NVARCHAR(length=100), Unicode(length=100)),
+        (DATE(), Date()),
+        (pg.JSON(), sa.JSON()),
+        (pg.ARRAY(sa.String), sa.ARRAY(sa.String)),
+    )
+    def test_as_generic(self, t1, t2):
+        assert repr(t1.as_generic()) == repr(t2)
+
+    @testing.combinations(*[(t,) for t in _all_types(omit_special_types=True)])
+    def test_as_generic_all_types(self, type_):
+        if issubclass(type_, ARRAY):
+            t1 = type_(String)
+        else:
+            t1 = type_()
+
```

Comment:
**mike bayer** (zzzeek) wrote:

what should as_generic() return when the argument is a TypeDecorator?  this is obviously not something we get back when we do reflection but if this is a 1st class method we should have some answer for this, even if it's NotImplementedError.   I think it would ideally be a TypeDecorator with an "impl" that's the generic type.

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2375